torch/library.py

```diff
@@ -245,6 +235,63 @@ def wrap(f):
     return wrap
 
 
+@functools.singledispatch
+def impl(name, dispatch_key, func=None, *, lib=None):
+    """Register an implementation for a DispatchKey for this operator.
+
+    torch.library.impl is a low-level API to directly register
+    an implementation for an operator for some DispatchKey in PyTorch.
+
+    This API may be used as a function or a decorator (see examples)
+
+    Args:
+        name (str): Should be a string that looks like "namespace::operator_name".
+        dispatch_key (str): The DispatchKey to register an impl for
+        lib (Optional[Library]): If provided, the lifetime of this registration
+            will be tied to the lifetime of the Library object.
+
+    Examples::
+        >>> import torch
+        >>> import numpy as np
+        >>>
+        >>> # Define the operator
+        >>> torch.library.define("mylibrary::sin", "(Tensor x) -> Tensor")
+        >>>
+        >>> # Add implementations for DispatchKey::CPU
+        >>> @torch.library.impl("mylibrary::sin", "CPU")
+        >>> def f(x):
+        >>>     return torch.from_numpy(np.sin(x.numpy()))
+        >>>
+        >>> x = torch.randn(3)
+        >>> y = torch.ops.mylibrary.sin(x)
+        >>> assert torch.allclose(y, x.sin())
+    """
+
+    def register(func):
+        namespace, _ = torch._library.utils.parse_namespace(name)
+        if lib is None:
+            use_lib = Library(namespace, "FRAGMENT")
+            _keep_alive.append(use_lib)
+        else:
+            use_lib = lib
```

Comment:
Do we want to add to `_registration_handles` like we do for `impl_abstract`?

Reply:
We need `_registration_handles` for impl_abstract because there is otherwise no way to delete an impl_abstract because it is not entirely managed by the PyTorch dispatcher.

All of the `impl` registrations are managed by the PyTorch dispatcher, so they do not need `_registration_handles`